numpy/core/src/multiarray/einsum_sumprod.c.src

```diff
@@ -10,38 +10,31 @@
 
 #define NPY_NO_DEPRECATED_API NPY_API_VERSION
 #define _MULTIARRAYMODULE
-
-#include <numpy/npy_common.h>
-#include <numpy/ndarraytypes.h>  /* for NPY_NTYPES */
 #include <numpy/halffloat.h>
-
 #include "einsum_sumprod.h"
 #include "einsum_debug.h"
+#include "simd/simd.h"
+#include "common.h"
 
-
-#ifdef NPY_HAVE_SSE_INTRINSICS
-#define EINSUM_USE_SSE1 1
+// ARM/Neon don't have instructions for aligned memory access
+#ifdef NPY_HAVE_NEON
+    #define EINSUM_IS_ALIGNED(x) 0
 #else
-#define EINSUM_USE_SSE1 0
-#endif
-
-#ifdef NPY_HAVE_SSE2_INTRINSICS
-#define EINSUM_USE_SSE2 1
-#else
-#define EINSUM_USE_SSE2 0
-#endif
-
-#if EINSUM_USE_SSE1
-#include <xmmintrin.h>
-#endif
-
-#if EINSUM_USE_SSE2
-#include <emmintrin.h>
-#endif
-
-#define EINSUM_IS_SSE_ALIGNED(x) ((((npy_intp)x)&0xf) == 0)
-
-/**********************************************/
+    #define EINSUM_IS_ALIGNED(x) npy_is_aligned(x, NPY_SIMD_WIDTH)
+#endif
+
+/**
+ * Unroll by four/eight scalars in case of:
+ *  - The main SIMD loop un-rolled by NPY_SIMD_WIDTH*(2|4), (2|4) represents
+ *    the number of times the loop that unrolled, Eg: for float32, 2 simd loop
+ *    is unrolled, for float64, 4 simd loop is unrolled, which may lead to
+ *    un-vectorize the remained scalars in bytes range <=NPY_SIMD_WIDTH*(2|4)
+ *    if The SIMD width is higher than 128bit, The performance loss on remained
+ *    arrays is nonnegligible, so we choose to use the compiler auto-vectorize.
+ *  - To give the chance to the compiler to
+ *    auto-vectorize in case of NPYV wasn't available.
+ */
```

Comment:
I'm still struggling to understand this comment, I think the language is slightly off. Happy to help with fixing it.

Am I right in thinking `@unroll_by@` is the `(2, 4)` here?

Reply:
Also, what is the difference between `NPY_SIMD` and `NPY_SIMD_WIDTH`? In this comment, you describe `NPY_SIMD` as the "SIMD width", which is confusing me.

Reply:
yes, when operating float array, @unroll_by@ equals 2, if operating double array, @unroll_by@ equals 4, NPY_SIMD represents how many bits can be hosted and NPY_SIMD_WIDTH represents how many bytes can be hosted, so NPY_SIMD is 8 times of NPY_SIMD_WIDTH. They represent the same thing with different alternate units, in this comment, SIMD width can described as 128bits or 16bytes.

Reply:
>  NPY_SIMD represents how many bits can be hosted and NPY_SIMD_WIDTH represents how many bytes can be hosted, so NPY_SIMD is 8 times of NPY_SIMD_WIDTH.

We already have precedent for naming these `NPY_BITSOF_FLOAT` and `NPY_SIZEOF_FLOAT`. Would it make sense (in a new PR) to rename `NPY_SIMD` &rarr; `NPY_SIZEOF_SIMD` and  `NPY_SIMD_WIDTH` &rarr; `NPY_BITSOF_SIMD` ?

Reply:
I think it's a good idea, The new name is more precise and clear, @seiko2plus what do you think?